File: parser/player_class.py
from __future__ import annotations

"""Détection de classe (calling) des joueurs Rift à partir des logs.

Objectif
--------
Associer un joueur à une classe (War / Rogue / Primalist / Cleric / Mage)
à partir de capacités *spécifiques à une classe*, repérées dans les logs.

On s'appuie sur des lignes comme :

  00:31:37: ( 6 , ... , Joueur , Joueur , 0 , ... , Inspirations profondes , 0 )
  Joueur acquiert la capacité Inspirations profondes de Joueur.

Le parser existant expose déjà :
- Event.code
- Event.src_kind / Event.src
- Event.ability

Ce module ne modifie pas le parsing : il ne fait que déduire une classe à partir
d'une table de correspondance "capacité -> classe".

Usage
-----
- Remplir ABILITY_TO_CLASS avec ~10 capacités par classe (en bas du fichier).
- Appeler infer_player_classes(events) pour obtenir un mapping joueur -> classe.

Si aucune capacité ne matche pour un joueur, la classe renvoyée est "Unknown"
(comme demandé).
"""

from collections import Counter, defaultdict
from typing import Dict, Iterable

from parser.types import Event

CLASSES = ("War", "Rogue", "Primalist", "Cleric", "Mage")
DEFAULT_CLASS = "Unknown"


def _norm(s: str) -> str:
    # Normalisation simple : casse + espaces
    return " ".join((s or "").strip().casefold().split())


# -----------------------------------------------------------------------------
# Capacité -> Classe
# -----------------------------------------------------------------------------

ABILITY_TO_CLASS: Dict[str, str] = {
    # Rogue
    #_norm(""): "Rogue",
    _norm("Mode gardien"): "Rogue",
    _norm("Maîtrise du combat à distance"): "Rogue",

    # War
    # _norm("..."): "War",
    _norm("Inspirations profondes"): "War",
    _norm("Retournement de lame"): "War",
    _norm("Néant"): "War",
    _norm("Posture prête"): "War",
    _norm("Conductivité améliorée"): "War",
    _norm("Lame-tempête"): "War",

    # Primalist
    # _norm("..."): "Primalist",
    _norm("Fontis de courroux"): "Primalist",
    _norm("Fontis de barbarie"): "Primalist",
    _norm("Fontis de sublimation"): "Primalist",
    _norm("Sauvagerie primitive"): "Primalist",
    _norm("Fontis d'oblitération"): "Primalist",
    _norm("Fontis de vitalité"): "Primalist",
    _norm("Fontis de soif de sang"): "Primalist",
    _norm("Fontis de rétribution"): "Primalist",
    _norm("Fontis de frénésie"): "Primalist",
    _norm("Destin en sursis"): "Primalist",
    _norm("Fontis de présence d'esprit"): "Primalist",

    # Cleric
    # _norm("..."): "Cleric",
    _norm("Bénédiction de vitalité"): "Cleric",
    _norm("Mine de souverain"): "Cleric",
    _norm("Armure de dévotion"): "Cleric",
    _norm("Armure du réveil"): "Cleric",
    _norm("Conduit de la Nature"): "Cleric",
    _norm("Vengeance du nord primitif"): "Cleric",
    _norm("Vengeance de la tempête hivernale"): "Cleric",
    _norm("Détermination héroïque"): "Cleric",
    _norm("Excès partagé"): "Cleric",

    # Mage
    # _norm("..."): "Mage",
    _norm("Rempart d'Archonte"): "Mage",
    _norm("Puissance de Neddra"): "Mage",
    _norm("Bienfait aqueux"): "Mage",
    _norm("Voile vivifiant"): "Mage",
    _norm("Armure prismatique"): "Mage",
    _norm("Charge foudroyante"): "Mage",
    _norm("Armure de Pyromancien"): "Mage",
    _norm("Voile des arcanes"): "Mage",
    _norm("Armure eldritch"): "Mage",
}


def infer_player_classes(events: Iterable[Event]) -> Dict[str, str]:
    """Retourne un mapping joueur -> classe.

    Stratégie :
    - on ne valide une classe que lorsqu'on voit un event "gain capacité" (code 6)
      dont le nom matche ABILITY_TO_CLASS.
    - si plusieurs classes matchent pour un même joueur (cas rare), on choisit
      la classe la plus fréquente (vote). En cas d'égalité, on renvoie Unknown.

    Notes :
    - on se base sur ev.src (le joueur qui acquiert)
    - on garde le code minimal et robuste, sans dépendre du texte "acquiert".
    """

    votes: dict[str, Counter[str]] = defaultdict(Counter)

    for ev in events:
        # Ligne "acquisition capacité" dans tes logs : code 6
        if ev.code != 6:
            continue
        if ev.src_kind != "P":
            continue
        if not ev.src:
            continue

        cls = ABILITY_TO_CLASS.get(_norm(ev.ability))
        if not cls:
            continue

        votes[ev.src][cls] += 1

    out: Dict[str, str] = {}

    for player, c in votes.items():
        if not c:
            out[player] = DEFAULT_CLASS
            continue

        most_common = c.most_common()
        if len(most_common) == 1:
            out[player] = most_common[0][0]
            continue

        # si égalité sur le max => Unknown
        top_cls, top_cnt = most_common[0]
        if any(cnt == top_cnt and cls != top_cls for cls, cnt in most_common[1:]):
            out[player] = DEFAULT_CLASS
        else:
            out[player] = top_cls

    return out

File: rust/src/player_class.rs
// src/player_class.rs
//
// Détection de classe (calling) des joueurs.
// Approche PT étendue : events code=6 (BuffGain) + table ability_name → classe.
// Deux noms par entrée (FR + EN) pour une couverture complète.

use std::collections::HashMap;
use once_cell::sync::Lazy;

use crate::types::{ActionCode, CombatEvent};

pub const DEFAULT_CLASS: &str = "Unknown";

fn norm(s: &str) -> String {
    s.trim().to_lowercase()
}

// ─────────────────────────────────────────────────────────────────────────────
// Table ability_name (normalisé) → classe
// FR + EN + DE pour chaque entrée
// ─────────────────────────────────────────────────────────────────────────────
static ABILITY_TO_CLASS: Lazy<HashMap<String, &'static str>> = Lazy::new(|| {
    let entries: &[(&str, &str)] = &[
        // ── Rogue ─────────────────────────────────────────────────────────
        ("mode gardien",                    "Rogue"),
        ("guardian phase",                  "Rogue"),
        ("maîtrise du combat à distance",   "Rogue"),
        ("ranged combat mastery",           "Rogue"),
        ("coda of wrath",                   "Rogue"),
        ("power chord",                     "Rogue"),
        ("guarded steel",                   "Rogue"),
        ("planar splash",                   "Rogue"),
        ("phantom blow",                    "Rogue"),
        ("rapid fire shot",                 "Rogue"),
        ("hellfire blades",                 "Rogue"),

        // ── Warrior ───────────────────────────────────────────────────────
        ("inspirations profondes",          "Warrior"),
        ("deep breaths",                    "Warrior"),
        ("retournement de lame",            "Warrior"),
        ("turn the blade",                  "Warrior"),
        ("néant",                           "Warrior"),
        ("void",                            "Warrior"),
        ("posture prête",                   "Warrior"),
        ("ready posture",                   "Warrior"),
        ("conductivité améliorée",          "Warrior"),
        ("enhanced conductivity",           "Warrior"),
        ("lame-tempête",                    "Warrior"),
        ("storm blade",                     "Warrior"),
        ("unstable reaction",               "Warrior"),
        ("tempest",                         "Warrior"),
        ("balance of power",                "Warrior"),
        ("protector's fury",                "Warrior"),
        ("positive reaction",               "Warrior"),
        ("mass casualty response",          "Warrior"),
        ("rising waterfall",                "Warrior"),

        // ── Primalist ─────────────────────────────────────────────────────
        ("fontis de courroux",              "Primalist"),
        ("font of wrath",                   "Primalist"),
        ("fontis de barbarie",              "Primalist"),
        ("font of savagery",                "Primalist"),
        ("fontis de sublimation",           "Primalist"),
        ("font of sublimation",             "Primalist"),
        ("sauvagerie primitive",            "Primalist"),
        ("primal savagery",                 "Primalist"),
        ("fontis d'oblitération",           "Primalist"),
        ("font of obliteration",            "Primalist"),
        ("fontis de vitalité",              "Primalist"),
        ("font of vitality",                "Primalist"),
        ("fontis de soif de sang",          "Primalist"),
        ("font of bloodlust",               "Primalist"),
        ("fontis de rétribution",           "Primalist"),
        ("font of retaliation",             "Primalist"),
        ("fontis de frénésie",              "Primalist"),
        ("font of frenzy",                  "Primalist"),
        ("destin en sursis",                "Primalist"),
        ("delayed fate",                    "Primalist"),
        ("fontis de présence d'esprit",     "Primalist"),
        ("font of wit",                     "Primalist"),
        ("crystalline smash",               "Primalist"),
        ("wild storms",                     "Primalist"),
        ("air lash",                        "Primalist"),
        ("tailwind",                        "Primalist"),

        // ── Cleric ────────────────────────────────────────────────────────
        ("bénédiction de vitalité",         "Cleric"),
        ("boon of vitality",                "Cleric"),
        ("mine de souverain",               "Cleric"),
        ("mien of leadership",              "Cleric"),
        ("armure de dévotion",              "Cleric"),
        ("armor of devotion",               "Cleric"),
        ("armure du réveil",                "Cleric"),
        ("armor of awakening",              "Cleric"),
        ("conduit de la nature",            "Cleric"),
        ("conduit of nature",               "Cleric"),
        ("vengeance du nord primitif",      "Cleric"),
        ("vengeance of the winter storm",   "Cleric"),
        ("vengeance de la tempête hivernale","Cleric"),
        ("vengeance of the primal north",   "Cleric"),
        ("détermination héroïque",          "Cleric"),
        ("heroic resolve",                  "Cleric"),
        ("excès partagé",                   "Cleric"),
        ("shared excess",                   "Cleric"),
        ("glacial insignia",                "Cleric"),
        ("wasting insignia",                "Cleric"),
        ("hammer of faith",                 "Cleric"),
        ("hammer of duty",                  "Cleric"),
        ("fae mimicry",                     "Cleric"),
        ("bound fate",                      "Cleric"),
        ("bolt of retribution",             "Cleric"),
        ("massive blow",                    "Cleric"),
        ("siphon vitality",                 "Cleric"),

        // ── Mage ──────────────────────────────────────────────────────────
        ("rempart d'archonte",              "Mage"),
        ("archont's bulwark",               "Mage"),
        ("archon's bulwark",                "Mage"),
        ("puissance de neddra",             "Mage"),
        ("neddra's might",                  "Mage"),
        ("bienfait aqueux",                 "Mage"),
        ("aqueous blessing",                "Mage"),
        ("voile vivifiant",                 "Mage"),
        ("lifegiving veil",                 "Mage"),
        ("armure prismatique",              "Mage"),
        ("prismatic armor",                 "Mage"),
        ("charge foudroyante",              "Mage"),
        ("lightning charge",                "Mage"),
        ("armure de pyromancien",           "Mage"),
        ("pyromancer's armor",              "Mage"),
        ("voile des arcanes",               "Mage"),
        ("arcane ward",                     "Mage"),
        ("armure eldritch",                 "Mage"),
        ("eldritch armor",                  "Mage"),
        ("burning purpose",                 "Mage"),
        ("granite salvo",                   "Mage"),
        ("counter shock",                   "Mage"),
        ("icy fury",                        "Mage"),
        ("shattered reflection",            "Mage"),
        ("ruin",                            "Mage"),
        ("condemn",                         "Mage"),
        ("atrophy",                         "Mage"),
    ];

    entries.iter()
        .map(|(name, class)| (norm(name), *class))
        .collect()
});

// ─────────────────────────────────────────────────────────────────────────────
// Inférence
// ─────────────────────────────────────────────────────────────────────────────
pub fn infer_player_classes(events: &[CombatEvent]) -> HashMap<String, String> {
    // votes[player] = {class → count}
    let mut votes: HashMap<String, HashMap<String, u32>> = HashMap::new();

    for ev in events {
        // Code 6 = BuffGain : acquisition d'une capacité ou aura (meilleur signal)
        if ev.code != ActionCode::BuffGain { continue; }
        if ev.src.is_empty()               { continue; }

        let key = norm(&ev.ability_name);
        if let Some(class) = ABILITY_TO_CLASS.get(&key) {
            *votes.entry(ev.src.clone())
                  .or_default()
                  .entry(class.to_string())
                  .or_insert(0) += 1;
        }
    }

    // Résolution : classe la plus fréquente ; égalité → Unknown
    let mut out = HashMap::new();
    for (player, counts) in votes {
        if counts.is_empty() {
            out.insert(player, DEFAULT_CLASS.to_string());
            continue;
        }
        let max_count = counts.values().copied().max().unwrap_or(0);
        let best: Vec<&String> = counts.iter()
            .filter(|(_, v)| **v == max_count)
            .map(|(k, _)| k)
            .collect();

        let cls = if best.len() == 1 {
            best[0].clone()
        } else {
            DEFAULT_CLASS.to_string()
        };
        out.insert(player, cls);
    }
    out
}
